node_helper.js

```diff
@@ -9,33 +10,34 @@ module.exports = NodeHelper.create({
 
 	async updateWOTrend (city, userAgent) {
 		let url = `https://www.wetteronline.de/wetter/${city}`;
-        let body = await this.getUrl(url, userAgent);
-        let gid = this.findGid(body);
+		let body = await this.getUrl(url, userAgent);
+		let gid = this.findGid(city, body);
 
-        const WO_DAILY_URL = `https://api-app.wetteronline.de/app/weather/forecast?av=2&mv=13&c=d2ViOmFxcnhwWDR3ZWJDSlRuWeb=&location_id=${gid}&timezone=${process.env.TZ}`;
-        let daily_promise = await this.getUrl(WO_DAILY_URL, userAgent);
-        let dailyData = JSON.parse(await daily_promise);
+		if(gid) {
+			const WO_DAILY_URL = `https://api-app.wetteronline.de/app/weather/forecast?av=2&mv=13&c=d2ViOmFxcnhwWDR3ZWJDSlRuWeb=&location_id=${gid}&timezone=${process.env.TZ}`;
+			let daily_promise = await this.getUrl(WO_DAILY_URL, userAgent);
+			let dailyData = JSON.parse(await daily_promise);
 
-        const WO_HOURLY_URL = `https://api-app.wetteronline.de/app/weather/hourcast?av=2&mv=13&c=d2ViOmFxcnhwWDR3ZWJDSlRuWeb=&location_id=${gid}&timezone=${process.env.TZ}`;
-        let hourly_promise = await this.getUrl(WO_HOURLY_URL, userAgent);
-        let hourlyData = JSON.parse(await hourly_promise);
+			const WO_HOURLY_URL = `https://api-app.wetteronline.de/app/weather/hourcast?av=2&mv=13&c=d2ViOmFxcnhwWDR3ZWJDSlRuWeb=&location_id=${gid}&timezone=${process.env.TZ}`;
+			let hourly_promise = await this.getUrl(WO_HOURLY_URL, userAgent);
+			let hourlyData = JSON.parse(await hourly_promise);
 
-		let event = this.extractEvent(dailyData, hourlyData, body);
-		
-		this.getHelper().sendSocketNotification("WETTERONLINE_RESULTS", event);
+			let event = this.extractEvent(dailyData, hourlyData, body);
+			
+			this.getHelper().sendSocketNotification("WETTERONLINE_RESULTS", event);
+		}
 	},
 
-	findGid (body) {
-		const exp = /WO\.geo = (\{[^\}]+\})/m;
-	
-		const match = body.match(exp);
+	findGid (city, body) {
+		const exp = /gid : "([^"]+)"/s;
+		const gid = body.match(exp);
 	
-		if(match) {
-			let obj = this.parseInlineJson(match[1]);
-			return obj["gid"];
+		if(gid) {
+			Log.info(`MMM-WetterOnline: The gid of city "${city}" is ${gid[1]}.`)
```

Comment:
nit-picking here, but wouldn't it be better if this method just *did one thing*™️ and the logging happened outside? seems superfluous to me to have the city parameter in this method just for logging something that the user supplies in config anyway... also a bit puzzled about the fact that we just don't return anything in the else branch and i assume `gid` outside is then undefined or something. is this common?

another question as a language foreigner: don't js people like to throw errors? if there's no gid, then clearly we might as well throw away the entire execution of `updateWOTrend()`. or is it a convention thing that we'd rather not let the async function throw an error and instead just log and execute gracefully?

Reply:
> wouldn't it be better if this method just did one thing™️ and the logging happened outside?

Surely this could be solved differently/better. For example, by adding a check whether a valid city name has been entered. With the PR, this check is now done by the findGid method. Before this PR only an error was thrown that was difficult for a user to understand.

> seems superfluous to me to have the city parameter in this method just for logging something that the user supplies in config anyway.

We can take the city out again if you like. But many MagicMirror users are not programmers and make a lot of mistakes when writing their configuration file. I can imagine that including the city can be helpful with this error message.

>  we just don't return anything in the else branch and i assume gid outside is then undefined or something. is this common?

It's not perfect, but it's a way of preventing throwing an error that doesn't help anyone.

Ejecting good error messages helps the user. Here is the difference if a user enter a wrong city name:

Before this PR:

```bash
[2024-06-16 12:14:06.660] [ERROR] (node:89014) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at JSON.parse (<anonymous>)
    at Class.updateWOTrend (/home/kristjan/MagicMirror/modules/MMM-WetterOnline/node_helper.js:19:25)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Class.socketNotificationReceived (/home/kristjan/MagicMirror/modules/MMM-WetterOnline/node_helper.js:7:4)
(Use `electron --trace-warnings ...` to show where the warning was created) 
[2024-06-16 12:14:06.661] [ERROR] (node:89014) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 7) 
```

After this PR:

```bash
[2024-06-16 12:15:16.356] [ERROR] MMM-WetterOnline: The gid of city "hedderts" could not be extracted. 
```

> don't js people like to throw errors?

In error handling, throwing errors is good. But that is only one part. The other part is catching them. Just throwing without catching is not proper error handling.

For many error cases with MagicMirror it is also sufficient/good practice to simply use `Log.error`. Writing software for a modular system like MagicMirror also makes it different from writing a standalone program. Especially, as mentioned above, non-programmers analyse the log output. Therefore, in my opinion, it makes sense to output typical errors with a clear error message.

Reply:
> But many MagicMirror users are not programmers and make a lot of mistakes when writing their configuration file.

fair point, i suppose. thanks for that perspective. since to me, i'm the primary target audience for this module, i guess i'm quick to forget about the average joe also using this.

> After this PR:

for the end user, i'm not sure that it's **that** much more understandable 🙂. but i agree, it looks cleaner. we'll keep it.

wrt errors: my main language is java and we like to throw exceptions where applicable. i just wondered how this paradigm might apply (or not) to JS development given your diff here. exception handling/error logging is then usually done in a method somewhere up the call stack where it makes sense to continue execution of that particular method. in this case, i guess i would've expected the framework (magic mirror) to somehow handle error catching/logging, given that this is interrupting an async method. idk how that works in JS though.

in any case, thanks for your patience in explaining these points.